File: cpp/problems/leetcode/0001_two_sum/solution.cpp
// LeetCode 1: Two Sum
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> seen;
        for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
            int need = target - nums[i];
            auto it = seen.find(need);
            if (it != seen.end())
                return {it->second, i};
            seen[nums[i]] = i;
        }
        return {};
    }
};

vector<int> two_sum_brute(vector<int>& nums, int target) {
    int n = static_cast<int>(nums.size());
    for (int i = 0; i < n; ++i)
        for (int j = i + 1; j < n; ++j)
            if (nums[i] + nums[j] == target)
                return {i, j};
    return {};
}

int main() {
    vector<int> nums{2, 7, 11, 15};
    auto ans = Solution().twoSum(nums, 9);
    assert(ans.size() == 2 && ans[0] == 0 && ans[1] == 1);
    cout << ans[0] << "," << ans[1] << endl; // 0,1

    auto brute = two_sum_brute(nums, 9);
    assert(brute == ans);
    return 0;
}

File: python/problems/leetcode/0001_two_sum/solution.py
"""LeetCode 1: Two Sum — 哈希表解法。"""


class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        seen: dict[int, int] = {}
        for i, x in enumerate(nums):
            need = target - x
            if need in seen:
                return [seen[need], i]
            seen[x] = i
        return []


def two_sum_brute(nums: list[int], target: int) -> list[int]:
    """暴力对照，仅用于理解。"""
    n = len(nums)
    for i in range(n):
        for j in range(i + 1, n):
            if nums[i] + nums[j] == target:
                return [i, j]
    return []


if __name__ == "__main__":
    assert Solution().twoSum([2, 7, 11, 15], 9) == [0, 1]
    assert two_sum_brute([2, 7, 11, 15], 9) == [0, 1]
    print(Solution().twoSum([2, 7, 11, 15], 9))  # [0, 1]
